File: Sem3/Advanced Methods of Programming/MyInterpreter - CSharp Console Application/MyInterpreter CSharp/domain/statement/AssignStatement.cs
using System;
using MyInterpreter_CSharp.domain.adt;
using MyInterpreter_CSharp.domain.expression;

namespace MyInterpreter_CSharp.domain.statement
{
    public class AssignStatement : IStatement
    {
        private string _id;
        private Expression _expression;

        public AssignStatement(string id, Expression expression)
        {
            _id = id;
            _expression = expression;
        }

        public ProgramState Execute(ProgramState programState)
        {
            IMyDictionary<string, int> symbolTable = programState.SymbolTable;
            try
            {
                int value = _expression.Eval(symbolTable);
                if (symbolTable.IsDefined(_id))
                    symbolTable.Update(_id, value);
                else
                    symbolTable.Add(_id, value);
            }
            catch (ExpressionException exception)
            {
                throw new StatementException(exception.Message + " " + _expression);
            }
            catch (AdtException exception)
            {
                throw new StatementException(exception.Message + " " + _id);
            }

            return null;
        }

        public override string ToString()
        {
            return _id + "=" + _expression;
        }

        public Expression Expression => _expression;
    }
}

File: Sem3/Advanced Methods of Programming/MyInterpreter - Java Console Application/project/src/domain/statement/AssignStatement.java
package domain.statement;


import domain.ProgramState;
import domain.adt.AdtException;
import domain.adt.MyDictionaryInterface;
import domain.expression.Expression;
import domain.expression.ExpressionException;


public class AssignStatement implements StatementInterface
{
    private String id;
    private Expression expression;


    public AssignStatement(String id, Expression expression)
    {
        this.id = id;
        this.expression = expression;
    }


    @Override
    public String toString()
    {
        return id + "=" + expression.toString();
    }


    @Override
    public ProgramState execute(ProgramState programState) throws StatementException
    {
        MyDictionaryInterface<String, Integer> symbolTable = programState.getSymbolTable();
        MyDictionaryInterface<Integer, Integer> heapTable = programState.getHeapTable();
        try
        {
            int val = expression.eval(symbolTable, heapTable);
            if (symbolTable.isDefined(id))
                symbolTable.update(id, val);
            else
                symbolTable.add(id, val);
        }
        catch (ExpressionException | AdtException exception)
        {
            throw new StatementException(exception.getMessage() + " Assign Statement.");
        }
        return null;
    }


    public Expression getExpression()
    {
        return expression;
    }
}
